「API > Composition API > Composition API: setup()」の翻訳 (#334)
vuejs-translations/docs-ja: src/api/composition-api-setup.md

# Composition API: setup()

:::info Note
このページでは、`setup` コンポーネントオプションの使用方法を説明します。単一ファイルコンポーネントで Composition API を使用する場合は、より簡潔で人間工学的な構文である [`<script setup>`](/api/sfc-script-setup.html) を使用することを推奨します。
:::

`setup()` フックは、次のような場合にコンポーネントで Composition API を使用するのためのエントリーポイントとして機能します:

1. ビルドステップなしでの Composition API の使用
2. Options API コンポーネント内の Composition API に基づくコードの統合

## 基本的な使い方

> 私たちは `setup()` からオブジェクトを返すことによって、 [Reactivity APIs](./reactivity-core.html) を使ったリアクティブな状態を宣言することや、それらを公開することができます。返されたオブジェクトのプロパティは、コンポーネントインスタンス上でも利用することができます (他のオプションが使用されている場合):

Reviewer: 1文目の by 以降は expose ~ の部分にかかるのではないかと思います
（また細かいですが、ここの主語はない方が自然かと思いました）

```suggestion
[Reactivity APIs](./reactivity-core.html) を使ってリアクティブな状態を宣言したり、`setup()` からオブジェクトを返すことによってそれらを公開することができます。返されたオブジェクトのプロパティは、コンポーネントインスタンス上でも利用することができます (他のオプションが使用されている場合):
```

</template>
```

> `setup` から返された [refs](/api/reactivity-core.html#ref) は、テンプレート内でアクセスされたときに[自動的に浅いアンラップ](/guide/essentials/reactivity-fundamentals.html#ref-unwrapping-in-templates)であるため、テンプレート内で `.value` を使用すべきではないことに注意してください。また、`this` でアクセスしたときも同様にアンラップです。

Reviewer: 「アンラップされる」とした方が元のニュアンスに近いかなと思いました
```suggestion
`setup` から返された [refs](/api/reactivity-core.html#ref) は、テンプレート内でアクセスされたときに[自動的に浅くアンラップされる](/guide/essentials/reactivity-fundamentals.html#ref-unwrapping-in-templates)ため、テンプレート内で `.value` を使用する必要はないことに注意してください。また、`this` でアクセスしたときも同様にアンラップされます。
```

:::tip

> `setup()` 自体はコンポーネントインスタンスにアクセスしません。- `this` は `setup()` 内の `null` 値を持ちます。Options API から Composition API で公開された値にアクセスすることができますが、その逆はできません。

Reviewer: こちらもニュアンスが少し変わっているので修正お願いします:pray:
```suggestion
`setup()` 自体はコンポーネントインスタンスにアクセスできません。- `this` は `setup()` 内では `null` 値を持ちます。Options API から Composition API で公開された値にアクセスすることができますが、その逆はできません。
```

}
```

もし、`props` オブジェクトを分割代入する場合は、分割代入された変数はリアクティビティを失うことに注意してください。 そのため、常に `props.xxx` の形で props にアクセスすることを推奨します。

もし、本当に props を分割代入すること、もしくはリアクティビティを保持しながら外部の関数に渡すことが必要なら、 ユーティリティー APIs である [toRefs()](./reactivity-utilities.html#torefs) や [toRef()](/api/reactivity-utilities.html#toref) を使用することで、行うことができます:

```js
import { toRefs } from 'vue'

export default {
  setup(props) {
    // `props` から refs オブジェクトへ変換し、分割代入を行います
    const { title } = toRefs(props)

>     // `title` は `props.title` を追跡する参照です

Reviewer: ```suggestion
    // `title` は `props.title` を追跡する ref です
```

    console.log(title.value)

    // もしくは、 `props` 内の単一プロパティから ref へ変換します
    const title = toRef(props, 'title')
  }
}
```

## Setup Context

`setup` 関数へ渡される第 2 引数は **Setup Context** オブジェクトです。context オブジェクトは `setup` 内で便利と思われる他の値を公開します:

```js
export default {
  setup(props, context) {
    // Attributes (非リアクティブオブジェクト、$attrs と同等です)
    console.log(context.attrs)

    // Slots (非リアクティブオブジェクト、$slots と同等です)
    console.log(context.slots)

    // Emit events (関数、$emit と同等です)
    console.log(context.emit)

    // パブリックプロパティの公開 (関数)
    console.log(context.expose)
  }
}
```

> context オブジェクトはリアクティブではなく、安全に分割代入されます:

Reviewer: ```suggestion
context オブジェクトはリアクティブではなく、安全に分割代入できます:
```

}
```

render 関数を返すことで、他のものを返すことができなくなります。内部的には問題ありませんが、このコンポーネントのメソッドをテンプレート参照から親コンポーネントに公開したい場合には、問題となります。

> 私たちは [`expose()`](#exposing-public-properties) を呼び出すことによって、この問題を解決することができます:

Reviewer: ```suggestion
[`expose()`](#exposing-public-properties) を呼び出すことによって、この問題を解決することができます:
```
